### PR TITLE
fix(planner): cap the row count of a unit CALL subquery

### LINKED ISSUE #2310: [Rust] Nested unit `CALL {}` subqueries do not cap cardinality, leaking inner fan-out
## Summary

A `CALL {}` subquery with no `RETURN` clause is a *unit subquery*: it must emit exactly one row per input row, regardless of how many rows it produces internally. On the Rust engine a **nested** unit subquery does not cap its cardinality, so the inner fan-out leaks into the outer row count.

Found while validating #624. The originally reported `@x not defined` error is gone on both engines, but the Rust engine produces wrong row counts for this query shape.

## Reproduction

```cypher
CREATE (:S), (:S), (:S)
```

```cypher
MATCH (x:S)
CALL {
  CALL { MATCH (y:S) RETURN y }
  CREATE (:Tmp)
}
RETURN count(*)
```

## Results

The outer `MATCH` yields 3 rows. The `CALL {}` block has no `RETURN`, so it is a unit subquery and must not change cardinality — the expected result is **3**.

| Engine | Result |
| --- | --- |
| `latest` (v4.20.1, C) | **3** ✅ |
| `edge-rs` (Rust) | **9** ❌ |

9 = 3 outer rows × 3 rows from the inner `MATCH (y:S)`, i.e. the inner subquery's cardinality propagates outward instead of being collapsed. With a third level of nesting this compounds to 27.

Uses named variables only — this is not related to the anonymous-relationship issue in #2307.

## Impact

Wrong row counts, and inflated side effects: the `CREATE` inside the subquery runs once per leaked row, so a query like this writes 9 nodes where it should write 3.

## Environment

* `falkordb/falkordb:edge-rs` — `graph` module ver `999999`, digest `sha256:82ce2fb13a1ae26e5adad723d01c6855942ade205c23a0ffb9f4ff5c9fd32ae1`
* `falkordb/falkordb:latest` — `graph` module ver `42001` (v4.20.1)

Related: #624, #2279

### PR DESCRIPTION
**Pull request by olifarhaan**
## TLDR

A `CALL {}` with no `RETURN` is a unit subquery: it runs for every input row and contributes no rows of its own. It was planned as `Apply(input, Optional(body))`, and `Optional` guarantees *at least* one row without capping at one, so the body's own fan-out multiplied the outer row count. Three `:S` nodes produced nine rows, and a third level of nesting twenty-seven

This gives it a dedicated operator that drains the body and then emits the input rows unchanged

Draining is the point. The writes are why the body exists, and a body that fans out internally performs them once per row it produced, so capping with a limit would have traded a wrong row count for a wrong write count

## Correcting the issue

The issue also reports the write count as wrong, expecting three `:Tmp` where nine are written. That part is not a bug. Measured against the C engine, which this repo treats as the reference:

| | rows | `:Tmp` written |
| --- | --- | --- |
| C (`falkordb-server:edge-c`) | 3 | 9 |
| Rust, before | 9 | 9 |
| Rust, after | 3 | 9 |

Only the row count ever differed. Nine writes is right: the body runs per input row, and the inner returning subquery yields three rows each time, so the `CREATE` fires nine times while the outer cardinality stays at three

## A body that collapses rows runs per row

Draining the whole input batch through one body instance is safe for the rows it produces, since they are all discarded, but not for its writes. A body that collapses rows applied once across the batch instead of once per input row, so its side effects happened the wrong number of times. Over three input rows:

| body | expected writes | batched | now |
| --- | --- | --- | --- |
| `WITH z LIMIT 1` | 3 | 1 | 3 |
| `WITH count(y) AS c` | 3 | 1 | 3 |
| `WITH z SKIP 1` | 3 | 5 | 3 |
| `WITH DISTINCT z` | 3 | 1 | 3 |

`SKIP` shows it most plainly: batched, it dropped one row out of six rather than one per invocation

`Apply` already draws this line for its own sub-plan, so the operator reuses the same predicate. A body containing an aggregation, `LIMIT`, `SKIP`, `DISTINCT`, `SORT`, `UNION`, `MERGE`, `OPTIONAL`, a cartesian product or a nested apply runs one row at a time, with the value dedupers cleared between invocations since each execution is its own `DISTINCT` scope. Bodies with nothing to collapse still run batched, so the fan-out this branch set out to contain stays contained

## User Flow

`GRAPH.QUERY` with a `CALL {}` that has no `RETURN` and whose body produces more than one row

## Relevant issues

Closes #2310

## Pre-Submission checklist

- [x] I have added meaningful tests
- [x] My PR passes all CI/CD checks (e.g., lint, format, unit tests)
- [x] My PR's scope is as isolated as possible; it only solves 1 specific problem

## Screenshots / Proof of Fix

Both engines live, side by side: this branch on 6399, the C reference on 6500

```bash
redis-server --port 6399 --loadmodule "$PWD/target/release/libfalkordb.dylib" &
docker run -d --platform linux/amd64 --name fdbc -p 6500:6379 \
  docker.io/falkordb/falkordb-server:edge-c

for p in 6399 6500; do
  redis-cli -p $p GRAPH.QUERY u "CREATE (:S), (:S), (:S)" > /dev/null
  echo "port $p"
  redis-cli -p $p GRAPH.QUERY u \
    "MATCH (x:S) CALL { CALL { MATCH (y:S) RETURN y } CREATE (:Tmp) } RETURN count(*)" | head -2
  redis-cli -p $p GRAPH.QUERY u "MATCH (t:Tmp) RETURN count(t)" | head -2
done
```

Before, on this branch's merge base:

```text
port 6399   rows: 9    Tmp: 9      <- outer cardinality multiplied by the body's fan-out
port 6500   rows: 3    Tmp: 9
```

After:

```text
port 6399   rows: 3    Tmp: 9
port 6500   rows: 3    Tmp: 9
```

The plan, which now names what it does instead of leaning on `Optional`:

```text
Aggregate
    Unit Subquery
        Node By Label Scan | (x:S)
        Commit
            Create | (_anon_0:Tmp),
                Apply
                    Argument
                    Project
                        Node By Label Scan | (y:S)
                            Argument
```

The C engine reaches the same shape with a dedicated `SubqueryForeach` operator

## Type

🐛 Bug Fix

## Changes

`graph/src/runtime/ops/unit_subquery.rs`, the new operator

`graph/src/planner/mod.rs`, the `IR::UnitSubquery` node, its display, and emitting it for a non-returning subquery. `ensure_apply_has_input` now covers it as well, since a query that opens on `CALL {}` has no clause to supply its input

`graph/src/runtime/{batch,runtime}.rs` and `ops/mod.rs`, registration and dispatch

`tests/flow/test_call_subquery.py`, one regression test over eight shapes

## QA runbook

The new test fails on the merge base and passes on this branch, with the fifty-three pre-existing tests in the file unaffected either way. Its assertions read as the bug:

```text
# merge base
Total Tests Run: 54, Total Tests Failed: 1, Total Tests Passed: 53
	[[9]] == [[3]]      body fans out to 3
	[[27]] == [[3]]     three levels of nesting
	[[6]] == [[3]]      UNWIND of 2 elements
	[[3]] == [[1]]      query opening on CALL {}
	18.0 == 12          two unit subqueries in sequence

# this branch
Total Tests Run: 54, Total Tests Failed: 0, Total Tests Passed: 54
```

Every shape was also diffed against the C engine directly, including bodies that write nothing, bodies that match nothing, `SET` and `DELETE` bodies, correlated bodies reading the outer variable, and two unit subqueries in sequence. Rows and `Nodes created` agree on all of them

One difference remains in that diff and is not related to this change: Rust reports `Labels added: 1` where C reports one per node. Plain `CREATE (:Tmp),(:Tmp),(:Tmp)` shows the same 1 against 3 with no subquery anywhere, so it is an engine-wide statistic difference

Full flow suite: `Total Tests Run: 1409, Total Tests Failed: 23, Total Tests Passed: 1386`, identical before and after. All 23 fail inside redis-py 8.1.0 with `Redis.__init__() got an unexpected keyword argument 'himport_registry'`, a client-side version mismatch in my local venv that cannot be reached from the module

TCK passes. `tests/test_e2e.py`, `tests/test_functions.py` and `tests/test_overflow_crashes.py`: 142 passed. `cargo test -p graph`: 116 passed. `cargo fmt --all --check` clean, `CXX=clang++ cargo clippy --all-targets` exits 0 leaving only the pre-existing `too many arguments` warning in `cond_var_len_traverse.rs`

Two things worth a reviewer's eye. The optimizer passes that special-case `IR::Apply` (`push_filters_down`, `fuse_optional_traverse`) do not match the new node, so nothing is pushed through a unit subquery. That is conservative and correct, but if pushing a filter past one is wanted it is a separate change. And `ApplyOp`'s `can_batch` list does not mention the new node either; every shape tested behaves, though a reviewer who knows that batching path may want to confirm the reasoning

### Final Attestation

- [x] The tests check the right things, including the edge cases, and regressions in the respective real-world customer use-cases are not possible after this PR

## Summary by CodeRabbit

* **Bug Fixes**
  * Corrected non-returning `CALL {}` subqueries so they preserve the number of outer rows.
  * Prevented inner subquery results from unintentionally multiplying outer query results.
  * Improved handling of nested subqueries, empty matches, `UNWIND`, sequential subqueries, implicit input rows, and write operations.
  * Ensured row-sensitive operations such as `LIMIT`, `SKIP`, `DISTINCT`, aggregation, and `MERGE` execute correctly within subqueries.

* **Tests**
  * Added coverage for row preservation, fan-out behavior, cleanup, nested execution, per-row processing, and execution plans involving non-returning subqueries.

### CODE REVIEW

**Comment by olifarhaan** [comment]
@coderabbitai review

Reverted the `///` comments to match each file's own convention, so docstring coverage will read lower here on purpose.

**Comment by AviAvni** [comment]
in master branch this call subquery foreach try to see if it is the same

File: graph/src/runtime/ops/unit_subquery.rs [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+
+        for row in batch.active_indices() {
+            let mut arg = BatchBuilder::new();
+            arg.push_row(&BatchRow::new(&batch, row).to_owned_row());
+            if let Err(e) = self.drain_body(arg.finish()) {
+                return Some(Err(e));
+            }
+            // Each invocation is a separate CALL {}, so DISTINCT inside the
+            // body must not remember rows from the previous one.
+            self.runtime.value_dedupers.borrow_mut().clear();
```
[the remark this answers is not in the document]
Comment (olifarhaan) [reply graph/src/runtime/ops/unit_subquery.rs]: Confirmed, and fixed in 3f974fcf6.

Filtering by plan subtree doesn't work here: the map is keyed by expression nodes, not plan nodes, so there's no subtree relation to filter on. The op now snapshots the key set once per input batch and retains only those keys after each invocation, so the body's entries drop while an enclosing `count(DISTINCT ...)` survives. A length check skips the pass when the body creates no dedup state

Worth flagging separately: `apply.rs:283` does the same `.clear()`, and `main` already returns 6 instead of 2 for `UNWIND range(1,3000) AS i WITH i, i % 2 AS p CALL { WITH i MATCH (y:S) WITH count(y) AS c RETURN c } RETURN count(DISTINCT p)`. Not touching it here, it needs its own fix
